src/main/scala/io/iohk/ethereum/blockchain/sync/regular/FetcherService.scala

```diff
@@ -0,0 +1,86 @@
+package io.iohk.ethereum.blockchain.sync.regular
+
+import akka.util.ByteString
+import cats.data.EitherT
+import io.iohk.ethereum.domain.{Block, BlockBody, BlockHeader, BlockchainReader}
+import io.iohk.ethereum.network.Peer
+import io.iohk.ethereum.network.p2p.messages.ETH62.{BlockBodies, BlockHeaders}
+import monix.eval.Task
+import io.iohk.ethereum.blockchain.sync.PeerRequestHandler.RequestFailed
+import io.iohk.ethereum.consensus.validators.BlockValidator
+import io.iohk.ethereum.utils.Config.SyncConfig
+import cats.implicits._
+
+import scala.annotation.tailrec
+
+//not used atm, a part of the future ExecutionSync
+class FetcherService(validator: BlockValidator, blockchainReader: BlockchainReader, syncConfig: SyncConfig) {
+  val batchSize = syncConfig.blockHeadersPerRequest
+
+  private def requestHeaders(
+      block: Either[BigInt, ByteString],
+      amount: BigInt
+  ): Task[Either[RequestFailed, BlockHeaders]] = ???
+
+  private def requestBodies(hashes: Seq[ByteString]): Task[Either[RequestFailed, BlockBodies]] = ???
+
+  private def requestStateNode(hash: ByteString): Task[Either[RequestFailed, Seq[ByteString]]] = ???
+
+  private def placeBlockInQueue(block: Block, peer: Peer): Unit = ???
+
+  def fetchBlocksUntil(
+      peer: Peer,
+      startFromBlock: Either[BigInt, ByteString],
+      fetchUntilBlock: Either[BigInt, ByteString]
+  ): EitherT[Task, RequestFailed, Unit] = {
+    val endNumber: Option[BigInt] =
+      fetchUntilBlock.fold(Some(_), blockchainReader.getBlockHeaderByHash(_).map(_.number))
+    val startNumber: Option[BigInt] =
+      startFromBlock.fold(Some(_), blockchainReader.getBlockHeaderByHash(_).map(_.number))
```

Comment:
correct me if I'm wrong but wouldn't it be beneficial to request a block by hash from a peer?

Reply:
right

Reply:
actually thinking more on that - we wouldn't know the hash of the block we want to fetch from the peer if we don't have that block ourselves. And the point of fetching from multiple peers is exactly to get alternative chains, so we do need to fetch by number here